tests/host_tools/fcmetrics.py

```diff
@@ -363,3 +367,126 @@ def validate_per_device_metrics(self, fc_metrics):
         if self.aggr_supported:
             metrics_aggregate = fc_metrics[self.dev_name]
             assert metrics_aggregate == metrics_calculated
+
+
+def get_emf_unit_for_fc_metrics(full_key):
+    """Returns CloudWatch Unit for requested FC metrics key"""
+    # We need to check each key because unit can be in group or key
+    # e.g.  latencies_us.diff_create_snapshot and
+    #       api_server.process_startup_time_us
+    for key in full_key.split("."):
+        if key.lower().endswith("_bytes") or key.lower().endswith("_bytes_count"):
+            return "Bytes"
+        if key.lower().endswith("_ms"):
+            return "Milliseconds"
+        if key.lower().endswith("_us"):
+            return "Microseconds"
+    return "Count"
+
+
+def flush_fc_metrics_to_cw(fc_metrics, metrics, assert_on_err=False):
+    """
+    Flush Firecracker metrics to CloudWatch
+    Use a new metrics logger and existing dimensions
+    Add a prefix "fc_metrics." to differentiate these metrics.
+    """
+    # List of SharedStoreMetric that once updated have the same value thoughout the life of vm
+    metrics_to_export_once = {
+        "api_server.process_startup_time_us",
+        "api_server.process_startup_time_cpu_us",
+        "latencies_us.full_create_snapshot",
+        "latencies_us.diff_create_snapshot",
+        "latencies_us.load_snapshot",
+        "latencies_us.pause_vm",
+        "latencies_us.resume_vm",
+        "latencies_us.vmm_full_create_snapshot",
+        "latencies_us.vmm_diff_create_snapshot",
+        "latencies_us.vmm_load_snapshot",
+        "latencies_us.vmm_pause_vm",
+        "latencies_us.vmm_resume_vm",
+    }
+    skip = set()
+    for group, keys in fc_metrics.items():
+        if group == "utc_timestamp_ms":
+            continue
+        for key, value in keys.items():
+            full_key = f"{group}.{key}"
+            # values are 0 when:
+            # - there is no update
+            # - device is not used
```

Comment:
I think 0s for "no update" should still be sent (i.e. this interface received 0 bytes in this period is still good info), the other ones we can skip. However in this function we cannot tell those two apart. If we also passed the Microvm object we could tell the difference, i.e. we could see if a device is configured in the VM and then send the metrics.

Reply:
I think microvm can provide list of disk and net devices attached but not other devices like vsock/balloon/entropy so uploading metrics based on devices seems difficult (unless I am missing something and there is a way to get the list ?). I agree that skipping metrics based on 0 doesn't look good either so removed that check. Now we'll upload metrics even if it is 0.

Reply:
Yeah that's probably simpler. 

For vsock and balloon I think we could probably get from the `/vm/config` api. But that's probably overkill.